### PR TITLE
Add Funnel and Multi-Analysis to the 'query' module

### PR DESCRIPTION
**Pull request by masojus**
This change adds Funnel and MultiAnalysis capabilities that work with the existing Query/KeenQueryClient functionality. 
- There is some refactoring to share more code between these different analysis types.
- This change is moving toward unifying large parts of MultiAnalysis and SingleAnalysis, the latter of which will replace Query. We can figure out how to best do that gently and without causing sudden/unnecessary churn in client code.
- Funnel and MultiAnalysis (and SingleAnalysis, which isn't used yet) have Builder objects that provide a more specific set of configuration options for their respective to-be-built classes. That means less guessing as a client dev and less grokking of the API reference to get the basic parameter set right. They also enforce more of the rules around parameters supported by the back end endpoints, so one knows at the build() call if something was configured wrong, rather than when that request is finally dispatched and the back end comes back with an error.
- Several tests were added for both Funnel and MultiAnalysis with combinations of required and optional parameters, but more can be added.
- README updated with changelog (guessed at the version) and samples.
- Travis build wasn't working with Android SDK 25.0 and Gradle Android plugin 2.2.0, so I played around with combinations of components and JDK in the .travis.yml, but if testing with JDK 6/7 is a requirement, we can look at changing that back while working through PR feedback.
- Ignore the Issue links in ebb2788 as those weren't intended to get pushed.

### CODE REVIEW

File: query/src/main/java/io/keen/client/java/MultiAnalysis.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+            return this;
+        }
+
+        /**
+         * Build the MultiAnalysis after arguments have been set.
+         *
+         * @return The new MultiAnalysis instance, configured and validated.
+         * @throws KeenQueryClientException if validating analysis parameters fails.
+         */
+        public MultiAnalysis build() throws KeenQueryClientException {
```
Comment (masojus) [inline query/src/main/java/io/keen/client/java/MultiAnalysis.java]: The throws clause can probably go away, and also the import statement.

File: query/src/main/java/io/keen/client/java/SingleAnalysis.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+            return this;
+        }
+
+        /**
+         * Build the SingleAnalysis after arguments have been set.
+         *
+         * @return The new SingleAnalysis instance, configured and validated.
+         * @throws KeenQueryClientException if validating analysis parameters fails.
+         */
+        public SingleAnalysis build() throws KeenQueryClientException {
```
Comment (masojus) [inline query/src/main/java/io/keen/client/java/SingleAnalysis.java]: Throws clause can probably go away.

File: query/src/test/java/io/keen/client/java/KeenQueryTest.java [hunk 3]
```diff
@@ -542,8 +559,1014 @@ public void testNullResult() throws Exception {
...
+
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        request.body.writeTo(outputStream);
+        String requestBody = outputStream.toString(ENCODING);
+        assertEquals(
+            "Unexpected request body",
+            "{\"steps\":[{\"timeframe\":\"this_7_days\",\"actor_property\":\"visitor.guid\",\"event_collection\":\"signed up\"},"
+          + "{\"timeframe\":\"this_7_days\",\"actor_property\":\"user.guid\",\"event_collection\":\"completed profile\"},"
+          + "{\"timeframe\":\"this_7_days\",\"timezone\":\"UTC\",\"actor_property\":\"user.guid\",\"event_collection\":\"referred user\"}]}",
+            requestBody);
```
Comment (baumatron) [inline query/src/test/java/io/keen/client/java/KeenQueryTest.java]: As discussed with Justin, I'd like to make sure this doesn't depend on the ordering of parameters that don't need to be ordered. Same goes for all other Funnel tests.

testAbsoluteTimeframe() also currently depends on the order of some unordered keys, so that should be updated as well.

File: query/src/main/java/io/keen/client/java/AbsoluteTimeframe.java [hunk 4]
```diff
@@ -46,16 +46,20 @@ public AbsoluteTimeframe(String start, String end) {
...
+
         if (start != null && end != null) {
-            Map absoluteTimeframe = new HashMap<String, Object>();
+            Map<String, Object> absoluteTimeframe = new HashMap<String, Object>(2);
+
             absoluteTimeframe.put(KeenQueryConstants.START, start);
             absoluteTimeframe.put(KeenQueryConstants.END, end);
 
-            timeframe = new HashMap<String, Object>();
+            timeframe = new HashMap<String, Object>(1);
```
Comment (baumatron) [inline query/src/main/java/io/keen/client/java/AbsoluteTimeframe.java]: Same here, I bet adding an item will cause re-allocation. Pretty minor but the purpose of specifying the initial allocation is lost.

File: query/src/main/java/io/keen/client/java/Filter.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    
+    // Required parameters
+    private final String propertyName;
+    private final FilterOperator operator;
+    private final Object propertyValue;
+    
+    public Filter(String propertyName, FilterOperator operator, Object propertyValue) {
+        
+        if (null == propertyName || propertyName.trim().isEmpty())
+        {
```
Comment (baumatron) [inline query/src/main/java/io/keen/client/java/Filter.java]: I messed up style consistency here with the opening curly braces.

File: query/src/main/java/io/keen/client/java/FunnelStep.java [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+     *                          of the query.
+     */
+    public FunnelStep(
+        final String collectionName,
+        final String actorPropertyName,
+        final Timeframe timeframe,
+        final Collection<Filter> filters,
+        final Boolean inverted,
+        final Boolean optional,
+        final Boolean withActors) {
```
Comment (baumatron) [inline query/src/main/java/io/keen/client/java/FunnelStep.java]: I think I'd like to try to clean this up to make it more readable from calling code. Either something with flags or Enums for special parameters I think would help.

File: .travis.yml [hunk 7]
```diff
@@ -1,18 +1,23 @@
...
+      - 'android-sdk-license-.+'
+      - 'google-gdk-license-.+'
 
 sudo: required
 
 addons:
   hostname: keenclient-java
 
 jdk:
-  - oraclejdk7
```
Comment (masojus) [inline .travis.yml]: We're hoping to turn back on jdk7 here.

File: query/src/main/java/io/keen/client/java/AbsoluteTimeframe.java [hunk 8]
```diff
@@ -46,16 +46,20 @@ public AbsoluteTimeframe(String start, String end) {
      */
     @Override
     public Map<String, Object> constructTimeframeArgs() {
-        Map timeframe = null;
+        Map<String, Object> timeframe = null;
+
         if (start != null && end != null) {
-            Map absoluteTimeframe = new HashMap<String, Object>();
+            Map<String, Object> absoluteTimeframe = new HashMap<String, Object>(2);
```
Comment (masojus) [inline query/src/main/java/io/keen/client/java/AbsoluteTimeframe.java]: Yeah oops. We'd want to divide by the default load factor I imagine if we wanted to keep this minor optimization. I'll change it in a follow-up commit.

File: query/src/main/java/io/keen/client/java/CollectionAnalysis.java [hunk 9]
```diff
@@ -0,0 +1,446 @@
...
+         *
+         * @param propertyName  The name of the property.
+         * @param operator      The operator (eg., gt, lt, exists, contains)
+         * @param propertyValue The property value. Refer to API documentation for info.
+         *                      This can be a string, number, boolean, or geo-coordinates
+         *                      and is based on what the operator is.
+         * @return This instance (for method chaining).
+         */
+        public ChainT withFilter(String propertyName,
+                                  FilterOperator operator,
```
Comment (masojus) [inline query/src/main/java/io/keen/client/java/CollectionAnalysis.java]: Strange spacing.

File: query/src/main/java/io/keen/client/java/Funnel.java [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+                }
+            }
+        }
+        
+        // Validate step properties that cannot be true for the first funnel step
+        FunnelStep firstStep = builder.steps.get(0);
+        
+        if (null != firstStep.getInverted() &&
+            true == firstStep.getInverted())
+        {
```
Comment (masojus) [inline query/src/main/java/io/keen/client/java/Funnel.java]: Nitpick: opening brace at EOL.

File: query/src/main/java/io/keen/client/java/Filter.java (hunk 5 above)
Comment (baumatron) [inline query/src/main/java/io/keen/client/java/Filter.java]: Fixed.

File: query/src/test/java/io/keen/client/java/KeenQueryTest.java (hunk 3 above)
Comment (baumatron) [inline query/src/test/java/io/keen/client/java/KeenQueryTest.java]: Fixed.

File: query/src/main/java/io/keen/client/java/Funnel.java (hunk 10 above)
Comment (baumatron) [inline query/src/main/java/io/keen/client/java/Funnel.java]: Fixed.

File: query/src/main/java/io/keen/client/java/KeenQueryRequest.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+import java.net.URL;
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * Interface to be implemented by a query request
+ *
+ * @author baumatron
+ */
+abstract class KeenQueryRequest {
```
Comment (Geeber) [inline query/src/main/java/io/keen/client/java/KeenQueryRequest.java]: Question: Why use an `abstract class` here instead of an actual `interface`? Is it just because of the default implementation in `getGroupByParams`? Doesn't necessarily seem worth it to me, but it's not a strong opinion.

File: query/src/main/java/io/keen/client/java/MultiAnalysis.java [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+            }
+        }
+
+        /**
+         * Adds a collection of sub-analyses.
+         *
+         * @param subAnalyses The new analysis arguments to add to the existing analyses.
+         * @return This instance (for method chaining).
+         */
+        public Builder withSubAnalyses(Collection<? extends SubAnalysis> subAnalyses) {
```
Comment (Geeber) [inline query/src/main/java/io/keen/client/java/MultiAnalysis.java]: Suggestion: Similar issue here as with the `setSteps`/`withSteps` in the funnel case. I think it generally improves API predictability if `withXxx(...)` can be assumed to always be implemented as `setXxx(...); return this;` (or `addXxx(...); return this;`, if that makes more sense semantically - like in the singular `withSubAnalysis` case below).

File: query/src/main/java/io/keen/client/java/MultiAnalysis.java [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+         *
+         * @param subAnalysis TThe sub-analysis to add to the existing analyses.
+         */
+        public void addSubAnalysis(SubAnalysis subAnalysis) {
+            if (null == subAnalysis) {
+                throw new IllegalArgumentException("The 'subAnalysis' parameter cannot be null.");
+            }
+
+            if (null == this.subAnalyses) {
+                // No point in specifying one twice. Though obviously for now, since SubAnalysis
```
Comment (Geeber) [inline query/src/main/java/io/keen/client/java/MultiAnalysis.java]: Question: Why is this comment here? I think I understand what the comment is saying, but it seems unrelated to the initialization of the `HashSet`...

File: query/src/main/java/io/keen/client/java/Percentile.java [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+     * @param percentile The percentile to calculate, supporting (0, 100.00] with two decimal places
+     *                   of precision.
+     * @return A Percentile instance representing the percentile to calculate.
+     */
+    public static Percentile createStrict(double percentile) {
+        if (MIN_PERCENTILE > percentile || MAX_PERCENTILE < percentile) {
+            throw new IllegalArgumentException("The 'percentile' must be in the range (0, 100].");
+        }
+
+        if (getNumDecimalPlaces(percentile) > 2) {
```
Comment (Geeber) [inline query/src/main/java/io/keen/client/java/Percentile.java]: Comment: I actually didn't know that we limit percentiles to 2 decimal places. Two and a half years here and I still learn new things about our API ;)

File: query/src/main/java/io/keen/client/java/Query.java [hunk 15]
```diff
@@ -151,38 +181,61 @@ protected Query(Builder builder) {
...
-         * @param filters  the filter arguments.
+         * Adds the given filters as optional parameters to the query.
+         *
+         * @param filters the filter arguments.
+         * @return This instance (for method chaining).
          */
-        public void setFilters(List<Map<String, Object>> filters) {this.filters = filters;}
-        public Builder withFilters(List<Map<String, Object>> filters) {
-            setFilters(filters);
+        public Builder withFilters(Collection<? extends Filter> filters) {
```
Comment (Geeber) [inline query/src/main/java/io/keen/client/java/Query.java]: Suggestion: Again, I would prefer to keep `withFilters` and `setFilters` consistent. (Note that this would also be a potentially backwards-compatibility-breaking change.)

File: query/src/test/java/io/keen/client/java/KeenQueryTest.java [hunk 16]
```diff
@@ -542,8 +612,1159 @@ public void testNullResult() throws Exception {
...
+            );
+            
+            funnelSteps.add(step);
+        }
+        
+        return funnelSteps;
+    }
+    
+    @Test
+    public void testFunnelWithOnlyRequiredParameters() throws Exception {
```
Comment (Geeber) [inline query/src/test/java/io/keen/client/java/KeenQueryTest.java]: Suggestion: I love to see all this test coverage! But I don't especially like seeing 1000+ line files, even if it's just tests. Might be worth breaking this up into a few separate files, e.g. one for funnel tests, one for multi-analysis tests, one for everything else (as a starting point). Not a ship-blocking requirement but I think it would make navigating through the tests a little easier.

File: query/src/main/java/io/keen/client/java/KeenQueryConstants.java [hunk 17]
```diff
@@ -3,10 +3,10 @@
 /**
  * Commonly used constants in the HTTP query request
  *
- * @author claireyoung
+ * @author claireyoung, baumatron, masojus
  * @since 1.0.0, 05/18/15
  */
-public class KeenQueryConstants {
+class KeenQueryConstants {
```
Comment (Geeber) [inline query/src/main/java/io/keen/client/java/KeenQueryConstants.java]: Comment: Changing this from `public` to default is, technically speaking, a breaking change (if somebody was, hypothetically, referencing constants from this class, their code would break). We can consider doing it anyway, but we should be careful about this kind of thing.

File: query/src/main/java/io/keen/client/java/Query.java [hunk 18]
```diff
@@ -36,8 +42,8 @@
      *
      * @return The JSON object map.
      */
-    public Map<String, Object> constructQueryArgs() {
-
+    @Override
+    Map<String, Object> constructRequestArgs() {
```
Comment (Geeber) [inline query/src/main/java/io/keen/client/java/Query.java]: Comment: Again, technically changing a `public` method to be non-public is a breaking change. In this case it's pretty hard to imagine anybody calling this in client code, but you never know...

File: query/src/main/java/io/keen/client/java/Query.java [hunk 19]
```diff
@@ -151,38 +181,61 @@ protected Query(Builder builder) {
...
         }
 
         /**
-         * get filters
-         * @return a list of filters.
+         * Get filters
+         *
+         * @return a collection of filters.
+         */
+        public Collection<Filter> getFilters() { return filters; }
```
Comment (Geeber) [inline query/src/main/java/io/keen/client/java/Query.java]: Concern: These changes to the filters-related methods are definitely breaking changes, and seem like they're actually pretty likely to break existing code. (Unless I'm missing something, which is very possible.)

**Comment by baumatron** [comment]
Thanks for the feedback, @Geeber! I'll look through a few of these comments now, though we may not have everything resolved until after the holidays.

In general, I understand the concern about backwards compatibility for a minor revision, and will work to update the change to preserve that.

Happy Turkey Day!

File: query/src/main/java/io/keen/client/java/AbsoluteTimeframe.java [hunk 20]
```diff
@@ -46,16 +46,20 @@ public AbsoluteTimeframe(String start, String end) {
      */
     @Override
     public Map<String, Object> constructTimeframeArgs() {
-        Map timeframe = null;
+        Map<String, Object> timeframe = null;
+
         if (start != null && end != null) {
-            Map absoluteTimeframe = new HashMap<String, Object>();
+            Map<String, Object> absoluteTimeframe = new HashMap<String, Object>(3);
```
Comment (baumatron) [inline query/src/main/java/io/keen/client/java/AbsoluteTimeframe.java]: I believe we're still compiling for source and target Java 6 compatibility. Maybe moving to something more recent could be added to a list of TODO's for the next major revision? Do you know what the most recent version is that could be targeted given business needs? Are there clients that require running on JRE 1.6? 1.7? Maybe there are implications for Android compatibility? Do you know why Java 6 was originally targeted?

File: query/src/main/java/io/keen/client/java/KeenQueryRequest.java (hunk 11 above)
Comment (baumatron) [inline query/src/main/java/io/keen/client/java/KeenQueryRequest.java]: This was done due to another Java limitation. I wanted to create a package internal interface. While interfaces themselves can have package-level visibility, methods are forced to have public visibility, so an object implementing an interface would be forced to make all interface methods public outside of the package, which isn't what we want.

Clients shouldn't need to know about any of these methods, probably shouldn't call them and if they do call them it'll have implications for them if we change the internal package object dependencies, something I'd argue is better to keep flexible. It would be better if the methods didn't pop up in auto-complete, too, cluttering things up and possibly making it harder to figure the SDK out. Unfortunately this has the side-effect that an object can only implement one of these "interfaces".

File: query/src/main/java/io/keen/client/java/Query.java (hunk 18 above)
Comment (baumatron) [inline query/src/main/java/io/keen/client/java/Query.java]: We were generally trying to wrangle in the exposed surface area of the SDK in the parts we touched that we deemed probably should have been package internal to begin with, but yeah a breaking change is a breaking change. Do you have a strong opinion one way or the other on this specific change? It would imply making this a public method on the KeenQueryRequest abstract base.

File: query/src/main/java/io/keen/client/java/KeenQueryConstants.java (hunk 17 above)
Comment (baumatron) [inline query/src/main/java/io/keen/client/java/KeenQueryConstants.java]: Again, we were trying to clean up visibility. Maybe we should plan for a package visibility audit for the next major revision instead?

File: README.md [hunk 21]
```diff
@@ -384,10 +384,87 @@ if (result.isIntervalResult()) {
     }
 }
 ```
+To perform a [Multi-Analysis](https://keen.io/docs/api/#multi-analysis), use the MultiAnalysis.Builder instead. An instance of MultiAnalysisResult will be returned unless Group By and/or Interval parameters are included, in which case the MultiAnalysisResult(s) will be nested inside an IntervalResult and/or a GroupByResult, just like any other QueryResult for other single analysis types when grouping/intervals are applied:
+
+``` java
+final MultiAnalysis multiAnalysis = new MultiAnalysis.Builder()
+        .withCollectionName("the_collection")
```
Comment (masojus) [inline README.md]: Sure. I made the case with @baumatron that if we were bumping major versions anyway, withEventCollection() sounds to me like it would expect something of type "Collection" or something similar, whereas withCollectionName() seems to imply it expects a String, but certainly I'm not adamant about that, and withEventCollection() matches the 'event_collection' JSON property name. Also, if using an IDE with intellisense, that's not really an issue--though those same IDEs will also likely have refactoring tools that make changing the method name two clicks. More importantly, we weren't really sure if we should shoot for a major version change or just a minor one (as evident with some of the other changes under discussion below). It'll get changed back.

File: query/src/main/java/io/keen/client/java/AbsoluteTimeframe.java (hunk 20 above)
Comment (masojus) [inline query/src/main/java/io/keen/client/java/AbsoluteTimeframe.java]: Yeah I've been meaning to ask about why we keep cross-compiling for 1.6 on all platforms across the board. We should have a deeper discussion about that, because this is just one of the ugly little retro things we have to keep dealing with if targeting 1.6.

File: query/src/main/java/io/keen/client/java/CollectionAnalysis.java [hunk 22]
```diff
@@ -0,0 +1,446 @@
...
+     * limited functionality.
+     *
+     * We could instead make these protected and force descendant classes to pick exactly which
+     * ones to make public, and therefore be in charge of returning the Builder subclass
+     * themselves, but for now they are all useful for subclasses.
+     *
+     * @param <ChainT> The type to return from fluent methods for chaining. Likely a derived builder
+     *                 type so as to continue chaining with the full capabilities of the subclass.
+     */
+    protected abstract static class Builder <ChainT> {
```
Comment (masojus) [inline query/src/main/java/io/keen/client/java/CollectionAnalysis.java]: Good eye. I'll address that as appropriate.

File: query/src/main/java/io/keen/client/java/KeenQueryRequest.java (hunk 11 above)
Comment (masojus) [inline query/src/main/java/io/keen/client/java/KeenQueryRequest.java]: Yeah, again we should do a follow-up push to limit the API surface of the SDK in general. I'd be all for that.

File: query/src/main/java/io/keen/client/java/Query.java (hunk 18 above)
Comment (masojus) [inline query/src/main/java/io/keen/client/java/Query.java]: Ditto. Similar to responses elsewhere, the fact that one can't imagine client code calling it implies it shouldn't really have been callable from another package to start with. So the question is this: how strictly we should adhere to semver here and in similar probably-shouldn't-have-been-that-way situations?

File: query/src/main/java/io/keen/client/java/Query.java (hunk 19 above)
Comment (masojus) [inline query/src/main/java/io/keen/client/java/Query.java]: Another case where we stepped half-way into the realm of major revision change, because Collection<> is more correct since List<> implies ordering, which I don't believe applies to filters. Also, Collection<Filter> seems more expressive and better hides internal details than WhateverCollection<Map<String, Object>> right? For a minor revision we can go back to the way it was.

File: query/src/main/java/io/keen/client/java/Percentile.java (hunk 14 above)
Comment (masojus) [inline query/src/main/java/io/keen/client/java/Percentile.java]: TheMoreYouKnow!

File: query/src/main/java/io/keen/client/java/Query.java (hunk 15 above)
Comment (masojus) [inline query/src/main/java/io/keen/client/java/Query.java]: So the issues is two-fold, here:
1) ```withFilters()``` now doesn't destroy existing filters
2) ```Collection<Filter-ish>``` as the parameter

In the second case, I'd say we can still be appropriately covariant and accept any Collection if we want to, and create a List from it...which is a step toward expressing filters as an unordered bag instead of a sequence. That shouldn't break existing code as long as T is  Map<String, Object> I believe. Thoughts?

File: query/src/test/java/io/keen/client/java/KeenQueryTest.java (hunk 16 above)
Comment (masojus) [inline query/src/test/java/io/keen/client/java/KeenQueryTest.java]: Second. I also think there are a couple hundred lines, at least, of copy/paste stuff in there that could really be trimmed down, but we didn't want to mess with existing tests until a later PR, at which point we can rip apart tests for different kinds of analyses and share utilities and such.

File: query/src/main/java/io/keen/client/java/KeenQueryConstants.java (hunk 17 above)
Comment (masojus) [inline query/src/main/java/io/keen/client/java/KeenQueryConstants.java]: I'm confused about this one, though. How could any client code be accessing these constants if they're all default access, which for class types is package-private, unless they wrote their code to be in this package, which is pretty suspect to begin with?

File: query/src/main/java/io/keen/client/java/MultiAnalysis.java (hunk 13 above)
Comment (masojus) [inline query/src/main/java/io/keen/client/java/MultiAnalysis.java]: I'm just saying that if you add the same exact analysis (and elsewhere filter also) we should throw because that's probably an error in client code, or else explicitly detect it and toss it since it'll just grow the payload of the request. HashSet, as opposed to other Collection types, imposes set semantics, so that just happens, but only if you override the hash/equality methods. I put the comment at the point where we pick the concrete type of the ```subAnalyses``` Collection since that's where the decision to use set semantics is ostensibly made. But it only really prevents us from duplicating instances with reference equality for the time being, hence the comment.

**Comment by masojus** [comment]
Thanks for the feedback, @Geeber ! I'll echo @baumatron 's sentiment that if we're targeting a minor revision, we can put some things back the way they were, and save the following sorts of things for a major revision:
1) Public API surface contraction.
2) More appropriate Collection type for filters and wherever else applicable, i.e. *List vs HashSet based on whether repetition is an error, List vs Collection based of whether ordering applies, LinkedList vs ArrayList based on whether indexed access is appropriate and common, and so forth.
3) ```withFilters()``` allowing fluent chaining no matter what you pass as a parameter in a less surprising manner.
4) Switch to ```SingleAnalysis``` as a replacement for ```Query```, which basically takes care of a bunch of 2) and 3)
5) Switch to Filter and in general not expose so much about the internal structure of things, e.g. - ```Map<String, Map<Object, <List<Map<String, Stuff>>>``` and so forth.
6) Refactor ```KeenQueryTest``` to share more code and not be thousands of lines all in one file. Break it out into different suites for different types of analyses, like single vs multi vs funnel at least. *This really could go into another PR for a minor revision unless we really want to do it for this PR.
7) Some other big feature additions.

File: query/src/main/java/io/keen/client/java/MultiAnalysis.java (hunk 12 above)
Comment (masojus) [inline query/src/main/java/io/keen/client/java/MultiAnalysis.java]: I'm with @baumatron on this one, and he and I of course discussed this a bit before going this route. It seemed more natural and more predictable to be able to interlace calls to ```withStep()``` and ```withSteps()``` at will, in any order, and any number of times with identical behavior. To me that's what with*() means for lists of things. For single properties, I think with*() should impose a one-time-only protocol, and we did that too, because why is your client code trying to with*() the collection name twice instead of calling set*()? But with the current design, withSteps() and withStep() can only be used either mutually-exclusively or in a very precise order to make any sense at all, which doesn't seem in the spirit of fluent APIs. To me, set*() is another beast entirely. Or are we thinking about that strangely?

Anyway, here since it's new code, we have the option to do it the way we all decide is best, or strictly adhere to how it used to be done for filters (those may be the same or different) for consistency, with the plan to potentially change it across the SDK in a major revision. Thoughts?

File: README.md (hunk 21 above)
Comment (josephwegner) [inline README.md]: I think `withCollectionName` is fine if we are doing a breaking version bump, as long as we use that everywhere.

File: query/src/main/java/io/keen/client/java/AbsoluteTimeframe.java (hunk 20 above)
Comment (josephwegner) [inline query/src/main/java/io/keen/client/java/AbsoluteTimeframe.java]: I'm fine with dropping support for Java 6. Official support for it ended a long time ago, and Android (the primary use case for this SDK) doesn't support it.

File: query/src/main/java/io/keen/client/java/Funnel.java [hunk 23]
```diff
@@ -0,0 +1,214 @@
...
+            }
+        }
+        
+        /**
+         * Set the list of funnel steps.
+         * 
+         * @param steps The funnel steps.
+         * @return This Builder instance with the steps added.
+         */
+        public Builder withSteps(List<? extends FunnelStep> steps) {
```
Comment (josephwegner) [inline query/src/main/java/io/keen/client/java/Funnel.java]: approved on punting this to a major version update.

File: query/src/main/java/io/keen/client/java/KeenQueryRequest.java (hunk 11 above)
Comment (josephwegner) [inline query/src/main/java/io/keen/client/java/KeenQueryRequest.java]: I think the abstract class is fine here, in order to limit the public surface.

File: query/src/main/java/io/keen/client/java/MultiAnalysis.java (hunk 12 above)
Comment (josephwegner) [inline query/src/main/java/io/keen/client/java/MultiAnalysis.java]: My primary concern is consistency with filters. As it's a breaking change for filters to adopt the new `setSteps`-like functionality, I say we also leave this as is for now.

File: query/src/main/java/io/keen/client/java/MultiAnalysis.java (hunk 13 above)
Comment (josephwegner) [inline query/src/main/java/io/keen/client/java/MultiAnalysis.java]: Maybe update the comment to make that clear. Right now it reads a bit jarringly for me as well. Prepending it with something like "Using a HashSet here, because it will only allow a single subAnalyses with the same name"

File: query/src/main/java/io/keen/client/java/AbsoluteTimeframe.java (hunk 20 above)
Comment (masojus) [inline query/src/main/java/io/keen/client/java/AbsoluteTimeframe.java]: Cool. We'll be sure to include some modernization to fully embrace JDK7 going forward.

File: query/src/main/java/io/keen/client/java/Funnel.java (hunk 23 above)
Comment (masojus) [inline query/src/main/java/io/keen/client/java/Funnel.java]: We are going to spin up a v6 branch internally and toss this sort of thing into it, then. Thanks!

File: query/src/main/java/io/keen/client/java/MultiAnalysis.java (hunk 13 above)
Comment (masojus) [inline query/src/main/java/io/keen/client/java/MultiAnalysis.java]: Will try to be clearer and more concise.
